Dynamic Library for iOS Carthage Support

lucasderraugh opened the pull request:
Trying to address issue #152, this build I believe will support Carthage. The issue is that Bolts is currently a static library, whereas Carthage only supports dynamic libraries (hence why the Mac build works). I know there can be backwards compatibility issues so I've added a new target that builds dynamically. I'd really like this to get moved into production so that Facebook could support Carthage out of the box. Let me know if anything needs to be changed.

nlutsenko commented:
Thanks for following up with this.
Let's address these few things, before we can merge it in:
- Our flow is to handle all target configurations in .xcconfig files (as you've seen) - can we extract all dynamic target specific things into a new .xcconfig file?
- Let's split changes around warnings and adding new target into 2 separate pull requests (I will send the warnings one in a bit), as they are not really relevant
- I guess the bundle id for the new target should remain the same as for static framework, since those are the same, just different way of packaging it (I don't feel strong about this, just a suggestion).

Let me know how I can help you with this.

nlutsenko commented:
@lucasderraugh, just opened #170 to fix all of these warnings and few more, no need to fix them here.

lucasderraugh commented:
@nlutsenko I think I've properly setup the config file for the dynamic library. I simply duplicated the existing iOS config and changed the MACH_O_TYPE to mh_dylib. What would you like me to do with the commit for warnings?

lucasderraugh commented:
I've made rebased with the changes @nlutsenko made to warnings and removed my commit for that. I've also squashed my commits into a single commit. Let me know if that's all that needs to be done to see this off.

jscubida commented:
Any update regarding this?

nlutsenko commented:
Looking now.

nlutsenko commented on Bolts.xcodeproj/project.pbxproj:
```diff
@@ -990,6 +1111,26 @@
+			isa = XCBuildConfiguration;
+			baseConfigurationReference = 1D88DFF31BD18B5600329FCD /* Bolts-iOS-Static.xcconfig */;
+			buildSettings = {
+				INFOPLIST_FILE = "$(SRCROOT)/Bolts/Resources/iOS-Info.plist";
```
Please kill :)
The whole point of having a xcconfig file is to avoid these.

nlutsenko commented on Configurations/Bolts-iOS-Static.xcconfig:
```diff
@@ -0,0 +1,24 @@
+DEFINES_MODULE = YES
+MODULEMAP_FILE = $(SRCROOT)/Bolts/Resources/iOS.modulemap
+
+OTHER_CFLAGS[sdk=iphoneos9.0] = $(inherited) -fembed-bitcode
```
`iphoneos9.*`, please :)

lucasderraugh replied:
I rebased on your changes for this.

nlutsenko commented:
Few more things here and there, overall looks great. Almost ready to ship.

jscubida commented:
Awesome thanks!

lucasderraugh commented:
Hopefully I deleted the right sections for this, let me know if there is anything else.

nlutsenko commented:
@lucasderraugh, looks great. Checked out things locally and here is the list of what we need to address before merging this. Let me know if there is anything I can help you with on any of these.

- Rename `Bolts-iOS-Static` to `Bolts-iOS-Dynamic`, since changing the default one to be dynamic will break people who depend on the packages that we distribute.
- Change mach-o type back to `staticlib` for the `Bolts-iOS` target
- Add `IPHONE_OS_DEPLOYMENT_TARGET` to `Bolts-iOS-Dynamic` and put 8.0 there, since dynamic frameworks are supported only starting from iOS `8.0`.
- Update project configuration and make `Bolts-iOS-Dynamic` target actually use the configuration file (it doesn't now)
- Reorder targets in xcodeproj and .xcconfig files in `Configurations` group in the project so both Bolts-iOS-* targets go one after each other
- ???
- PROFIT!

lucasderraugh commented:
Sorry for the commit amends, wasn't sure if you strive for nice git history or not (from the looks of it you do). Regardless, I believe all these changes have been made (including a rebase).

lucasderraugh commented:
Ok, so in the Bolts-iOS-Dynamic config file, I added the CoreGraphics and UIKit framework for linker params. This for whatever reason is breaking the OSX tests. Not sure how to fix this by using the config files exclusively.

nlutsenko commented on Configurations/Bolts-iOS.xcconfig:
```diff
@@ -11,7 +11,6 @@
 #include "Shared/Product/Framework.xcconfig"
 
 PRODUCT_NAME = Bolts
-IPHONEOS_DEPLOYMENT_TARGET = 6.0
```
I believe this was unintended change, please revert it.

nlutsenko commented:
@lucasderraugh 
> ammends

It's ok to keep separate commits until we are fully done, then we can squash everything together.
Might be easier to track progress this way, but absolutely up to you.

> frameworks and OS X tests breakage

Not sure why it got broken. Try rebasing it against the current latest master.
Also, I've noticed that for some reason the xcodeproj changes included renaming of the existing Bolts-iOS target.
If nothing helps at this point - my suggestion is to try to reset all the changes for the Xcode project and recreate the target from scratch (by duplicating the existing one + renaming + changing the xcconfig file attached to this).


Takes a while to get through all of these, but I am confident we are going to make it together!

lucasderraugh commented:
Ok, I just completely rebased off of what you had and redid what I originally set out to do.

lucasderraugh commented:
Could this get merged in?

nlutsenko commented:
Doing last sanity check right now and merging in...

nlutsenko commented:
@lucasderraugh Found a small nit issue - `Bolts-iOS-Dynamic.xcconfig` file was by mistake added to `BoltsTests-iOS` target as resource. Please remove it (simply uncheckmark the target in the Utilities pane) and i'll merge it in straight away.

lucasderraugh commented:
@nlutsenko Should be good from here.

nlutsenko commented:
Perfect! One last thing - could you squash both commits into just 1?

lucasderraugh commented:
Done

nlutsenko commented:
Thanks for bearing with me through the whole review. So glad it is finally done!
